### PR TITLE
go: add CoreFoundation and Security frameworks for cgo on darwin

### PR DESCRIPTION
**Pull request by bouk**
Here I'm adding some CGO flags to ensure the c compiler is able to find system frameworks on Darwin. These frameworks need to be always available, to ensure some parts of the standard library (notably `crypto/x509`) can be compiled.

I've also added a test that reproduces the issue.

###### Motivation for this change

See https://github.com/NixOS/nixpkgs/issues/56348 for background

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/go/1.14.nix [hunk 1]
```diff
@@ -229,6 +232,17 @@ stdenv.mkDerivation rec {
     runHook postInstall
   '';
 
+  # These flags are needed to ensure the standard library compiles correctly
+  postInstall = optionalString stdenv.isDarwin (with darwin.apple_sdk.frameworks; ''
+    wrapProgram $out/share/go/bin/go \
+      --suffix CGO_CFLAGS  ' ' '-iframework ${CoreFoundation}/Library/Frameworks -iframework ${Security}/Library/Frameworks' \
+      --suffix CGO_LDFLAGS ' ' '-F${CoreFoundation}/Library/Frameworks -F${Security}/Library/Frameworks'
```
Comment (bouk) [inline pkgs/development/compilers/go/1.14.nix:239]: I could also use NIX_CFLAGS and NIX_LDFLAGS, not sure what's best?

**Comment by kalbasit** [comment]
I have not tried this yet, but I can confirm that this is a constant source of annoyance for us because Go crashes anytime we import something. On my end (at work) I have this snippet which works (copied only relevant parts, but you get the idea):

```
  # in a let statement
  appleFrameworks = with pkgs.darwin.apple_sdk.frameworks; [
    "-F${CoreFoundation}/Library/Frameworks"
    "-F${CoreServices}/Library/Frameworks"
    "-F${Foundation}/Library/Frameworks"
    "-F${Security}/Library/Frameworks"
  ];

        CGO_CFLAGS = optionalString isDarwin
          (builtins.concatStringsSep " " (builtins.filter (e: e != "") ([
            (builtins.getEnv "CGO_CFLAGS") ] ++ appleFrameworks)));
        CGO_LDFLAGS = builtins.concatStringsSep " " (builtins.filter (e: e != "") (
          [ (builtins.getEnv "CGO_LDFLAGS") ]
          ++ optionals isDarwin (["-lstdc++" "-lm" ] ++ appleFrameworks)

```

I'm not sure if we need all frameworks though.

**Comment by bouk** [comment]
I looked through the go source, CoreFoundation and Security is all that's needed

**Comment by Mic92** [comment]
We already do in `go`:

```
depsTargetTargetPropagated = optionals stdenv.isDarwin [ Security Foundation ];
```

This means in a `nix-shell` there will be both frameworks presents.
Does go pickup our `NIX_LDFLAGS`/`NIX_CFLAGS_COMPILE`?
If not rather than wrapping the go compiler we could add a build support hook in go that also sets `CGO_CFLAGS` and `CGO_LDFLAGS`.

**Comment by Mic92** [comment]
Also cc @LnL7 and @Ericson2314

**Comment by bouk** [comment]
Yep, it works with nix-shell. Under the hood cgo uses $CC.

But what I (and others) are running against is that you can't use the compiler directly outside of a nix-shell, since those NIX_LDFLAGS are not set. 

Another solution might be for nix-Darwin et al to add flags that are set through depsTargetTargetPropagated (and similar things) in the environment, so the user's shell picks them up. I'd personally prefer the wrapper approach since it reduces the scope of where the flags get set.

**Comment by bouk** [comment]
Closing in favor of https://github.com/NixOS/nixpkgs/pull/91347

**Pull request closed without merging**